python/preemo/worker/_messaging_client.py

```diff
@@ -1,61 +1,50 @@
+from preemo.gen.services.worker_pb2_grpc import WorkerServiceStub
 
 
 @runtime_checkable
 class IMessagingClient(Protocol):
-    def send_worker_request(self, worker_request: WorkerRequest) -> WorkerReply:
+    def register_function(
+        self, request: RegisterFunctionRequest
+    ) -> RegisterFunctionResponse:
         pass
 
 
-# Instantiating this class will hang if it cannot connect to a valid server
 class MessagingClient:
     def __init__(self, *, worker_server_url: str) -> None:
-        context = zmq.Context()
-
-        # TODO(adrian@preemo.io, 02/25/2023): investigate other socket types, such as PUSH/PULL
-        self._socket = context.socket(zmq.REQ)
-        # TODO(adrian@preemo.io, 02/15/2023): add logging indicating attempting to connect and successful connection
-        # TODO(adrian@preemo.io, 02/15/2023): add timeout, like Promise.race (python threads)
-        # TODO(adrian@preemo.io, 02/16/2023): investigate which, if any of these timeouts are useful
-        # self._socket.setsockopt(zmq.CONNECT_TIMEOUT, 1_000)
-        # self._socket.setsockopt(zmq.SNDTIMEO, 1_000)
-        # self._socket.setsockopt(zmq.RCVTIMEO, 1_000)
-        self._socket.connect(worker_server_url)
-
-        header_reply = self._send_header_request(HeaderRequest(version=__version__))
-        if header_reply.status != STATUS_OK:
+        self._channel = grpc.insecure_channel(target=worker_server_url)
+        self._worker_service = WorkerServiceStub(self._channel)
+
+        header_response = self._initiate(HeaderRequest(version=__version__))
+        if header_response.status != STATUS_OK:
             raise Exception(
-                f"worker server replied to header request with unexpected status: {header_reply.status} and message: {header_reply.message}"
+                f"worker server replied to header request with unexpected status: {header_response.status} and message: {header_response.message}"
             )
 
-    def _send_message(self, message: bytes) -> bytes:
-        self._socket.send(message)
-        return self._socket.recv()
-
-    def _send_header_request(self, header_request: HeaderRequest) -> HeaderReply:
-        message = header_request.SerializeToString()
-        reply = self._send_message(message)
-
-        return HeaderReply.FromString(reply)
-
-    def send_worker_request(self, worker_request: WorkerRequest) -> WorkerReply:
-        message = worker_request.SerializeToString()
-        reply = self._send_message(message)
+    def _initiate(self, request: HeaderRequest) -> HeaderResponse:
+        return self._worker_service.Initiate(request)  # type: ignore
 
-        return WorkerReply.FromString(reply)
+    def register_function(
+        self, request: RegisterFunctionRequest
+    ) -> RegisterFunctionResponse:
+        # TODO(adrian@preemo.io, 03/03/2023): can we do better with the types?
+        return self._worker_service.RegisterFunction(request)  # type: ignore
```

Comment:
did a bit of looking and came upon [a protoc plugin for python](https://github.com/danielgtaylor/python-betterproto) not sure that this solves the problem, but it might be worth checking out

Reply:
I attempted to use betterproto and I'm going to summarize my findings here.

Pros:
- The generated files were easier to read and included types. No need for pyi files.
- The package field in the proto files was respected.
- The imports were created correctly and i no longer needed to mess with sys.path.

Cons:
- The current stable version does not support the optional keyword in proto3 so we would have to use the beta version.
- For some still unknown reason I was unable to use `--custom_opt` in order to use pydantic models instead of dataclasses.
- We found a few github issues indicating that betterproto was at least an order of magnitude slower than protoc.
- The grpc service was only generated for async. It seems there were a handful of ways we could've gotten that to work, but we cut off that avenue due to the other issues.